### DISCUSSION TITLE
How to skip package if the fixture for the package fails to initialize data.

### DISCUSSION THREAD

**Discussion by sukeshlaghate** [pytest-dev/pytest#9747] [Q&A]
My pyTest suite is organized in following structure

            TestSuite
                |
                |-- conftest.py
                |
                |--package1
                |      |
                |      |--__init__.py
                |      |
                |      |--fixtures
                |      |      |
                |      |      |--data_setup1.py
                |      |
                |      |-- test_pack_1_cycle_1.py
                |            ...
                |
                |--package2
                |      |
                |      |--__init__.py
                |      |
                |      |--fixtures
                |      |      |
                |      |      |--data_setup2.py
                |      |
                |      |-- test_pack_2_cycle_1.py
                |                ...
                |-- ...
now suppose that fixture data_setup2 fails to initialize data required by all tests contained in package2, then all tests under package2 should be skipped. While all other tests belonging to other packages should run.

What I have tried so far:
**Approach 1:**

1.  In conftest.py I have following code

```python
  pytest_plugins = [
          fixture.replace(os.path.sep, ".").replace("fixtures.py", "fixtures") for fixture in iglob("**/*fixtures.py", recursive=True)
   ]
```
this allows me to collect all fixtures

2. in individual data_setup*.py i have defined fixutre as

```python
  @pytest.fixture(scope="package", autouse=True)
  def setup_data_2(establish_http_connection):
      print('\n calling package level fixture')
      # do the setup
      if not check_flag:
          pytest.skip("skiping package2 due to failure in setting up data")
      else:
          yield data_packet
          # do tearing up of data_packet
```                
However, failure in one of the data setup causes all the packages to be skipped. What I would expect is that pytest should skip only those tests that belong to package2

Approach2:
I define fixtures and have check flag returned as part of data_packet and in individual test_pack_2_cycles_*.py I am checking to see if the success flag is present
like so

1. In conftest.py I have following code

 ```python
 pytest_plugins = [
    fixture.replace(os.path.sep, ".").replace("fixtures.py", "fixtures") for fixture in iglob("**/*fixtures.py", recursive=True)
]
```
this allows me to collect all fixtures
2. in data_setup*.py I have defined fixture as

```python
@pytest.fixture(scope="package", autouse=True)
def setup_data_2(establish_http_connection):
    print('\n calling package level fixture')
    # do the setup
    data_packet = {
        "success_flag" : True  # if data setup succeeded
        ... # data related to setup
    }
    yield data_packet
    # do data packet tearing
```
3. In individual test_pack_2_cycle_*.py (for e.g in test_pack_2_cycle_1.py) i have:

```python
 class TestPack2Cycle1():
    @fixture(autouse=True)
    def _setup(self, establish_http_connection, setup_data):
        print('TestPack2Cycle1 Setup is ', setup_data['success_flat'])
        if not setup_data['success_flag']:
            pytest.skip('Skiping as setup data failed', allow_module_level=True)
```
This works as expected, however this approach requires me to write these lines in every test__cycle inside the package. Is there a better approach to skip tests for package where data setup (fixture belonging to particular package) fails? 
I would like to implement approach 1 or something similar on that lines and thus avoid having to write pytest.skip in individual *.py files.

Request you to provide how I can achieve this or point me in the right direction.

With regards
Sukesh.

**Comment by nicoddemus** [comment]
Hi @sukeshlaghate,

Unfortunately seems like a bug related to package scoped fixtures: in your example, they are only being setup once per session, instead of once per package. 😕
